samcli/commands/local/lib/sam_api_provider.py

```diff
@@ -159,55 +170,68 @@ def _convert_event_api(lambda_logical_id, event_properties):
-            List of APIs obtained by combining both the input lists.
+        list of samcli.local.apigw.local_apigw_service.Route
+            List of routes obtained by combining both the input lists.
         """
 
-        implicit_apis = []
-        explicit_apis = []
+        implicit_routes = []
+        explicit_routes = []
 
         # Store implicit and explicit APIs separately in order to merge them later in the correct order
         # Implicit APIs are defined on a resource with logicalID ServerlessRestApi
         for logical_id, apis in collector:
             if logical_id == SamApiProvider.IMPLICIT_API_RESOURCE_ID:
-                implicit_apis.extend(apis)
+                implicit_routes.extend(apis)
             else:
-                explicit_apis.extend(apis)
+                explicit_routes.extend(apis)
 
         # We will use "path+method" combination as key to this dictionary and store the Api config for this combination.
         # If an path+method combo already exists, then overwrite it if and only if this is an implicit API
-        all_apis = {}
+        all_routes = {}
 
         # By adding implicit APIs to the end of the list, they will be iterated last. If a configuration was already
         # written by explicit API, it will be overriden by implicit API, just by virtue of order of iteration.
-        all_configs = explicit_apis + implicit_apis
+        all_configs = explicit_routes + implicit_routes
 
         for config in all_configs:
             # Normalize the methods before de-duping to allow an ANY method in implicit API to override a regular HTTP
-            # method on explicit API.
-            for normalized_method in AbstractApiProvider.normalize_http_methods(config.method):
+            # method on explicit route.
+            for normalized_method in config.methods:
                 key = config.path + normalized_method
-                all_apis[key] = config
+                all_routes[key] = config
 
-        result = set(all_apis.values())  # Assign to a set() to de-dupe
+        result = set(all_routes.values())  # Assign to a set() to de-dupe
         LOG.debug("Removed duplicates from '%d' Explicit APIs and '%d' Implicit APIs to produce '%d' APIs",
-                  len(explicit_apis), len(implicit_apis), len(result))
+                  len(explicit_routes), len(implicit_routes), len(result))
+
+        # Merge route methods so that the combo of path and function_name is unique
+        # This is accomplished by combining the methods
+        merged_route_methods = {}
+        for route in result:
+            key = route.function_name + route.path
+            if key in merged_route_methods:
+                new_methods = list(set(merged_route_methods[key].methods + route.methods))
+                merged_route_methods[key].methods = new_methods
+            else:
+                merged_route_methods[key] = route
 
+        result = list(merged_route_methods.values())
```

Comment:
Not sure this belongs here. This method is to combine implicit and explicit apis. This changes the scope of the funciton. While this isn't wrong, it doesn't make the code as readable as it could. For example, we already do this function name + path when we print out routes in local_api_service.py. We should find a better how to do this deduping.